File: src/main/java/com/udalmik/shop/promo/ThirdFreePromo.java
package com.udalmik.shop.promo;

import com.udalmik.shop.model.Purchase;

import java.math.BigDecimal;

public class ThirdFreePromo implements ItemPromo {
    @Override
    public BigDecimal getDiscount(Purchase purchase) {
        final var price = purchase.getItem().getPrice();
        final var quantity = purchase.getQuantity();
        final var freeItemsCount = quantity / 3;
        return price.multiply(BigDecimal.valueOf(freeItemsCount));
    }
}

File: src/main/java/com/udalmik/shop/promo/ThirdFreePromo.kt
package com.udalmik.shop.promo

import com.udalmik.shop.model.Purchase
import java.math.BigDecimal

class ThirdFreePromo : ItemPromo {
    override fun getDiscount(purchase: Purchase): BigDecimal {
        val price = purchase.item.price
        val quantity = purchase.quantity
        val freeItemsCount = quantity / 3
        return price.multiply(BigDecimal.valueOf(freeItemsCount))
    }
}